File: 백준/Silver/1012. 유기농 배추/유기농 배추.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {
    static int M, N, K;
    static int[][] field;
    static boolean[][] visited;
    static int[] dx = {-1, 1, 0, 0};
    static int[] dy = {0, 0, -1, 1};

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();

        int T = Integer.parseInt(br.readLine());

        while (T-- > 0) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            M = Integer.parseInt(st.nextToken());
            N = Integer.parseInt(st.nextToken());
            K = Integer.parseInt(st.nextToken());

            field = new int[N][M];
            visited = new boolean[N][M];

            for (int i = 0; i < K; i++) {
                st = new StringTokenizer(br.readLine());
                int x = Integer.parseInt(st.nextToken());
                int y = Integer.parseInt(st.nextToken());
                field[y][x] = 1;
            }

            int worms = 0;

            for (int i = 0; i < N; i++) {
                for (int j = 0; j < M; j++) {
                    if (field[i][j] == 1 && !visited[i][j]) {
                        bfs(i, j);
                        worms++;
                    }
                }
            }

            sb.append(worms).append("\n");
        }

        System.out.print(sb);
    }

    private static void bfs(int y, int x) {
        Queue<int[]> queue = new LinkedList<>();
        queue.offer(new int[]{y, x});
        visited[y][x] = true;

        while (!queue.isEmpty()) {
            int[] current = queue.poll();
            int cx = current[1];
            int cy = current[0];

            for (int d = 0; d < 4; d++) {
                int nx = cx + dx[d];
                int ny = cy + dy[d];

                if (nx >= 0 && nx < M && ny >= 0 && ny < N &&
                        field[ny][nx] == 1 && !visited[ny][nx]) {
                    visited[ny][nx] = true;
                    queue.offer(new int[]{ny, nx});
                }
            }
        }
    }
}

File: 백준/Silver/1012. 유기농 배추/유기농 배추.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*

lateinit var vis: Array<BooleanArray>
lateinit var board: Array<IntArray>
val dx = arrayOf(1, 0, -1, 0)
val dy = arrayOf(0, 1, 0, -1)

fun main() = with(BufferedReader(InputStreamReader(System.`in`))) {
    val t = readln().toInt()
    repeat(t) {
        val (m, n, k) = readln().split(" ").map { it.toInt() }
        vis = Array(n) { BooleanArray(m) }
        board = Array(n) { IntArray(m) }
        val queue: Queue<Pair<Int, Int>> = LinkedList()
        repeat(k) {
            val (a, b) = readln().split(" ").map { it.toInt() }
            board[b][a] = 1
        }

        var cnt = 0
        repeat(n) { col ->
            repeat(m) { row ->
                if (board[col][row] == 1 && !vis[col][row]) {
                    vis[col][row] = true
                    queue.add(Pair(col, row))
                    while (queue.isNotEmpty()) {
                        val cur = queue.first()
                        queue.poll()
                        for (dir in 0 until 4) {
                            val nx = cur.first + dx[dir]
                            val ny = cur.second + dy[dir]
                            if (nx < 0 || nx >= n || ny < 0 || ny >= m) continue
                            if (board[nx][ny] != 1 || vis[nx][ny]) continue
                            vis[nx][ny] = true
                            queue.add(Pair(nx, ny))
                        }
                    }
                    cnt++
                }
            }
        }
        println(cnt)
    }

}